Backward compatability tests for rolling upgrades

abeyad opened the pull request:
This PR introduces backward compatibility index tests to test the rolling upgrade process amongst Elasticsearch instances within the same major version.  The test executes in three phases.  In the first phase, we form a cluster of 2 ES instances on an old version.  In the second phase, we keep one of the nodes from the old cluster, kill the other node, but preserve its data directory and start an instance of the current version of ES using the same data directory as the killed instance.  In the third phase, we kill the other old version ES instance from the first phase and launch a new instance, using the same data directory as the killed instance.  Therefore, during phase 3, we have fully migrated and have all current versions of ES running.  In each phase, we run REST tests that index documents and search them, ensuring at each stage that the documents from the previous phase are still there.

Note that because we haven't released a GA yet of 5.0, the tests currently don't start an old version cluster in the first phase.  Once GA is released, this will be changed to make the backward compatibility version 5.0, while the current version in the cluster will be 5.x.

rjernst committed:
Add rolling upgrade test project

Commit:
Adds tests for rolling upgrades to execute

rjernst committed:
Made node config always have a unicast transport uri closure

rjernst committed:
Merge branch 'master' into rolling_upgrades

rjernst committed:
Use 5.0 alpha5 for bwc version, so we have transport.ports

Also fixed bug to ensure unicast host is writtent in yaml quotes

Commit:
Changed rest-api-spec tests folder structure

Commit:
Reworking yaml tests for rolling upgrades

Commit:
Added a indices flush to the rolling upgrades rest tests to ensure
they are on disk for the upgraded nodes to pick up.

Commit:
Prevent the rolling upgrades rest tests from cleaning up indices
after finishing if a the tests.rest.preserve_indices system property
is set

Commit:
Better checks for the cluster being up in the rolling upgrades tests.

Commit:
When Elasticsearch nodes are started in gradle to form a
cluster, we wait for the cluster health to indicate the
necessary nodes have formed a cluster.  This check was an
exact value (equality) check.  However, if we are trying to
connect the nodes in the cluster to nodes from a previously
formed cluster (of the same name), then we will have more
nodes returned by the cluster health check than the current
task's configured number of nodes.  Hence, this check needs
to be a >= check.  This commit fixes it.

Commit:
In the rolling upgrades tests, we do not want to stop nodes
automatically between tasks, as we want some of the nodes from
the previous task to continue running in the next task. This
commit enables a cluster configuration setting to not stop
nodes automatically after a task runs, but instead the creator
of the test task must stop the running nodes explicitly in a
cleanup phase.

Commit:
Merge remote-tracking branch 'upstream/master' into rolling_upgrades

Commit:
Disables unit tests for rolling upgrades, as there are only
rest integration tests.

Commit:
Removes stopNodeUponCompletion in favor of moving the stop
nodes task to the final part of the cluster test task execution
graph.

nik9000 commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterFormationTasks.groovy:
```diff
@@ -48,7 +48,7 @@ class ClusterFormationTasks {
      *
      * Returns a NodeInfo object for the first node in the cluster.
```
Out of data javadoc.

abeyad replied:
I pushed https://github.com/elastic/elasticsearch/pull/20532/commits/98230d035a8dd1e8ae0a82058db73a9f8d03d66e

nik9000 commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterFormationTasks.groovy:
```diff
@@ -95,22 +95,15 @@ class ClusterFormationTasks {
         Task wait = configureWaitTask("${task.name}#wait", project, nodes, startTasks)
         task.dependsOn(wait)
 
         // delay the resolution of the uri by wrapping in a closure, so it is not used until read for tests
```
This comment also seems out of date? Maybe it just needs to be moved.

abeyad replied:
I pushed https://github.com/elastic/elasticsearch/pull/20532/commits/98230d035a8dd1e8ae0a82058db73a9f8d03d66e, it was outdated based on the new unicastTransportUri closure

nik9000 commented on test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java:
```diff
@@ -139,13 +139,16 @@ protected final RestClient adminClient() {
-            // 404 here just means we had no indexes
-            if (e.getResponse().getStatusLine().getStatusCode() != 404) {
-                throw e;
+        final boolean preserveIndices = Boolean.parseBoolean(System.getProperty("tests.rest.preserve_indices"));
```
I feel like this should be a method you override and return `true` from if you want to preserve the indices from the test. That way it can be Javadoced and you can use your IDE to see who overrides it.

abeyad replied:
I pushed https://github.com/elastic/elasticsearch/pull/20532/commits/98230d035a8dd1e8ae0a82058db73a9f8d03d66e

nik9000 commented on qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/UpgradeClusterClientYamlTestSuiteIT.java:
```diff
@@ -0,0 +1,43 @@
+@TimeoutSuite(millis = 40 * TimeUnits.MINUTE) // some of the windows test VMs are slow as hell
+public class UpgradeClusterClientYamlTestSuiteIT extends ESClientYamlSuiteTestCase {
+
+    public UpgradeClusterClientYamlTestSuiteIT(ClientYamlTestCandidate testCandidate) {
```
Now that we have `ESRestTestCase` you can write the test in java if you like. It is up to you if you want to do so, but I like having stuff like loops available. Yaml is nice if you want to make sure the language clients can run the tests but that really only applies to the tests in `rest-api-spec`.

abeyad replied:
I started to go down this path.  I normally would think this is a much better plan because of all the extra logic at your disposal with Java as opposed to the yaml files.  But in this case, since all I was doing was making REST calls and making sure they succeeded, it turned out the Yaml made executing the requests and parsing the responses much more straightforward than writing the Java logic to create the request body as json, and read the response stream and construct a json parser and parse the response.  Since I wasn't doing much more than that in these tests, I felt it made more sense to keep them as yaml tests here.

Commit:
Adds a preserveIndicesUponCompletion method to ESRestTestCase
that can be overridden by subclasses if the test must not
delete indices it created after exiting.

nik9000 commented on qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/UpgradeClusterClientYamlTestSuiteIT.java:
```diff
@@ -0,0 +1,48 @@
+
+import java.io.IOException;
+
+@TimeoutSuite(millis = 40 * TimeUnits.MINUTE) // some of the windows test VMs are slow as hell
```
Do we need this on this one? It seems like these test suites are very small and any of them taking 40 minutes is grounds for making someone look at the VM at a minimum.

abeyad replied:
I will set this timeout to 5 mins, anything over would seem worthy of someone's attention.

abeyad commented:
@nik9000 thanks for reviewing!

Commit:
Change the timeout of the rolling upgrades test from 40 mins to 5 mins
to still allow accounting for slow VMs

abeyad commented:
5.0 commit: 2bc19131d49714a8670b7949c19d069d9646a5a0
5.x commit: ad198245fa387a582a520e19eebd5a14ba7e5e69
